supabase/migrations/50_controllers.sql

```diff
@@ -0,0 +1,127 @@
+  'Represents the type of a dependency of one spec on another. This enum is a
+  strict superset of catalog_spec_type, for historical reasons.';
+
+-- This works because the `flow_type` enum is a superset of the `catalog_spec_type` enum.
+-- This approach was taken from:
+-- https://www.munderwood.ca/index.php/2015/05/28/altering-postgresql-columns-from-one-enum-to-another/
+alter table live_spec_flows
+alter column flow_type set data type flow_type
+using flow_type::text::flow_type;
+
+-- We now allow live specs to be deleted even though they're still referenced by other specs.
+-- This means we need to relax the foreign key constraints on `live_spec_flows`.
+alter table live_spec_flows
+drop constraint live_spec_flows_source_id_fkey;
+alter table live_spec_flows
+drop constraint live_spec_flows_target_id_fkey;
+
+alter table live_specs add column controller_next_run timestamptz;
+comment on column live_specs.controller_next_run is 'The next time the controller for this spec should run.';
+
+-- Create a partial covering index on live specs to facilitate querying for the next controller run.
+-- This is used by `deque` in `agent-sql/src/controllers.rs`.
+create index live_specs_controller_next_run on live_specs(controller_next_run)
+include (id)
+where controller_next_run is not null;
+
+-- This constraint is removed because we're changing how we represent deleted specs, so that only
+-- the `spec` column is null. Setting `spec_type` to null was unnecessary, and retaining it is
+-- now necessary in order for `live_spec_flows` to stay consistent with `live_specs` in case of
+-- spec deletions that don't draft all the connected specs.
+alter table live_specs drop constraint "spec and spec_type must be consistent";
+alter table draft_specs drop constraint "spec and spec_type must be consistent";
+
+-- The goal is to make `spec_type` non-nullable. There's a bunch of existing soft-deleted
+-- live_specs that already have a null `spec_type`, though, so we we need to either delete those
+-- rows or set the `spec_type` to a non-null value. Deleting them could _technically_ make it
+-- possible for old agent instances to re-create `live_specs` with the same name during the rollout
+-- period, so we instead set the type to `test`. The test controller will run and hard-delete the
+-- `live_specs` row. Using `test` means that we won't try to modify any shards or journals as part
+-- of that process.
+update live_specs set spec_type = 'test' where spec_type is null;
+alter table live_specs alter column spec_type set not null;
+
+-- Allow spec_type to remain non-null for deleted specs
+alter table publication_specs drop constraint "spec and spec_type must be consistent";
+
+
+create table controller_jobs (
+    -- The name of the live spec that this pertains to
+    live_spec_id flowid not null references live_specs (id) on delete cascade,
+    -- The version of the controller that last updated this row. Used to identify controllers to run
```

Comment:
🤔 how does this impact query performance of the controller? If we're querying for rows where the controller version is less than our current version, that would need to be indexed or it'll require a full table scan (i haven't gotten to this code yet; or is it an output-only column?).

If the goal is to be able to trigger controllers on software releases, it seems like we could as-easily do that through a migration that tickles `controller_next_run` on `live_specs` -- that might even be better because we could set up a schedule that staggers / slow-rolls the controller runs.

Reply:
We talked about this on VC and [this commit](https://github.com/estuary/flow/pull/1489/commits/c33c583b4242cfdc3afa6870052aafe9d4bd4dea) introduces a better approach for the migration. Also edited the pr description to add an overall deployment plan.